### PR TITLE
FIX: pandas indexing error

### LINKED ISSUE #5550: Plotting datetime values from Pandas dataframe
This appears to be a new issue in 1.5.0.

The script below attempts to plot two 2-D graphs whose X and Y values are Pandas series.  The issue seems to occur when pyplot is passed a datetime column which doesn't contain an index of value 0 - note how the second dataframe contains only odd indices (1, 3, 5, etc.)

```
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# create sample dataframe with preset dates and values columns
dates = np.arange('2005-02', '2005-03', dtype='datetime64[D]')
values = np.sin(np.array(range(len(dates))))
df = pd.DataFrame({'dates': dates, 'values': values})

# matplotlib figure + two subplots for comparison
fig, axes = plt.subplots(1, 2)

# create two dataframes for comparison - one with all indices, including 0, and one with only odd indices
with_zero_index = df.copy()
without_zero_index = df[np.array(df.index) % 2 == 1].copy()

# plot both - note how second plot fails without a 0 index
axes[0].plot(with_zero_index['dates'], with_zero_index['values'])
axes[1].plot(without_zero_index['dates'], without_zero_index['values'])
```

Stack trace:
```
KeyError                                  Traceback (most recent call last)
<ipython-input-29-28e878247c17> in <module>()
     17 # plot both - note how second plot fails without a 0 index
     18 axes[0].plot(with_zero_index['dates'], with_zero_index['values'])
---> 19 axes[1].plot(without_zero_index['dates'], without_zero_index['values'])

/home/jim/arcemweb/venv/local/lib/python2.7/site-packages/matplotlib/__init__.pyc in inner(ax, *args, **kwargs)
   1809                     warnings.warn(msg % (label_namer, func.__name__),
   1810                                   RuntimeWarning, stacklevel=2)
-> 1811             return func(ax, *args, **kwargs)
   1812         pre_doc = inner.__doc__
   1813         if pre_doc is None:

/home/jim/arcemweb/venv/local/lib/python2.7/site-packages/matplotlib/axes/_axes.pyc in plot(self, *args, **kwargs)
   1425             kwargs['color'] = c
   1426 
-> 1427         for line in self._get_lines(*args, **kwargs):
   1428             self.add_line(line)
   1429             lines.append(line)

/home/jim/arcemweb/venv/local/lib/python2.7/site-packages/matplotlib/axes/_base.pyc in _grab_next_args(self, *args, **kwargs)
    384                 return
    385             if len(remaining) <= 3:
--> 386                 for seg in self._plot_args(remaining, kwargs):
    387                     yield seg
    388                 return

/home/jim/arcemweb/venv/local/lib/python2.7/site-packages/matplotlib/axes/_base.pyc in _plot_args(self, tup, kwargs)
    362             x, y = index_of(tup[-1])
    363 
--> 364         x, y = self._xy_from_xy(x, y)
    365 
    366         if self.command == 'plot':

/home/jim/arcemweb/venv/local/lib/python2.7/site-packages/matplotlib/axes/_base.pyc in _xy_from_xy(self, x, y)
    195     def _xy_from_xy(self, x, y):
    196         if self.axes.xaxis is not None and self.axes.yaxis is not None:
--> 197             bx = self.axes.xaxis.update_units(x)
    198             by = self.axes.yaxis.update_units(y)
    199 

/home/jim/arcemweb/venv/local/lib/python2.7/site-packages/matplotlib/axis.pyc in update_units(self, data)
   1387         neednew = self.converter != converter
   1388         self.converter = converter
-> 1389         default = self.converter.default_units(data, self)
   1390         if default is not None and self.units is None:
   1391             self.set_units(default)

/home/jim/arcemweb/venv/local/lib/python2.7/site-packages/matplotlib/dates.pyc in default_units(x, axis)
   1562 
   1563         try:
-> 1564             x = x[0]
   1565         except (TypeError, IndexError):
   1566             pass

/home/jim/arcemweb/venv/local/lib/python2.7/site-packages/pandas/core/series.pyc in __getitem__(self, key)
    519     def __getitem__(self, key):
    520         try:
--> 521             result = self.index.get_value(self, key)
    522 
    523             if not np.isscalar(result):

/home/jim/arcemweb/venv/local/lib/python2.7/site-packages/pandas/core/index.pyc in get_value(self, series, key)
   1593 
   1594         try:
-> 1595             return self._engine.get_value(s, k)
   1596         except KeyError as e1:
   1597             if len(self) > 0 and self.inferred_type in ['integer','boolean']:

pandas/index.pyx in pandas.index.IndexEngine.get_value (pandas/index.c:3113)()

pandas/index.pyx in pandas.index.IndexEngine.get_value (pandas/index.c:2844)()

pandas/index.pyx in pandas.index.IndexEngine.get_loc (pandas/index.c:3704)()

pandas/hashtable.pyx in pandas.hashtable.Int64HashTable.get_item (pandas/hashtable.c:7224)()

pandas/hashtable.pyx in pandas.hashtable.Int64HashTable.get_item (pandas/hashtable.c:7162)()

KeyError: 0
```

Looks like matplotlib's dates.py module is attempting to access the first value in the datetime series - but when passed as a pandas series, x[0] represents the value at index 0, and a) might not exist, and b) isn't necessarily the first value in the series!  Possible fix might be to catch IndexErrors and attempt an x.iloc[0] call instead.

Script above worked fine in 1.4.3, but fails specifically in 1.5.0.

### PR DESCRIPTION
**Pull request by tacaswell**
pd.Series prefer indexing via searching the index to positional
indexing.  This method will get the first element of any iterable.

It will advance a generater, but they did not work previously anyway.

Closes #5550

### CODE REVIEW

File: lib/matplotlib/dates.py [hunk 1]
```diff
@@ -1561,7 +1561,7 @@ def default_units(x, axis):
             x = x.ravel()
 
         try:
-            x = x[0]
+            x = next(iter(x))
         except (TypeError, IndexError):
```
Comment (has2k1) [inline lib/matplotlib/dates.py]: line `1565` should be modified to

```python
except (TypeError, StopIteration):
```

Comment (tacaswell) [inline lib/matplotlib/dates.py]: Thanks for catching that!

**Comment by tacaswell** [comment]
rebased.

**Comment by tacaswell** [comment]
Also addresses #5557 , tests coming.

File: lib/matplotlib/axes/_axes.py [hunk 2]
```diff
@@ -5871,6 +5871,40 @@ def hist(self, x, bins=10, range=None, normed=False, weights=None,
...
         """
+        def _normalize_input(inp, ename='input'):
+            """Normalize 1 or 2d input into list of np.ndarray or
+            a single 2D np.ndarray.
+
+            Parameters
+            ----------
+            inp : iterable
+            ename : str, optional
+                Name to use is ValueError if can not be normalized
```
Comment (WeatherGod) [inline lib/matplotlib/axes/_axes.py]: "is" --> "in"?

Comment (tacaswell) [inline lib/matplotlib/axes/_axes.py]: I swear I know english...

File: lib/matplotlib/axes/_axes.py [hunk 3]
```diff
@@ -5871,6 +5871,40 @@ def hist(self, x, bins=10, range=None, normed=False, weights=None,
         .. plot:: mpl_examples/statistics/histogram_demo_features.py
 
         """
+        def _normalize_input(inp, ename='input'):
+            """Normalize 1 or 2d input into list of np.ndarray or
+            a single 2D np.ndarray.
```
Comment (WeatherGod) [inline lib/matplotlib/axes/_axes.py]: So, which does it do? A list of ndarrays or a 2D array? This isn't very clear.

**Comment by WeatherGod** [comment]
besides the nitpicking, this looks good to me

File: lib/matplotlib/axes/_axes.py (hunk 3 above)
Comment (tacaswell) [inline lib/matplotlib/axes/_axes.py]: it really does return either/or :imp: 

Note that this is a `_`'d private to this method function.  The main reason for this existing is to remove copy-paste code so the questionable API is OK (imo) as it preserves the behavior of the block of code the function call is replacing.

**Comment by tacaswell** [comment]
@WeatherGod do you want me to drop that (some what superfluous) white space commit?

File: lib/matplotlib/axes/_axes.py (hunk 3 above)
Comment (WeatherGod) [inline lib/matplotlib/axes/_axes.py]: Right, the API is awkward, but I am fine with it as it is private and internal, and it does clean up the code (and fixes a possible mutation side-effect in weights processing). I was more aiming for a clearer docstring for future devs. The if-statements are quite opaque.

**Comment by WeatherGod** [comment]
Eh, might be better to get rid of that commented out code entirely. I don't see the point of keeping it around.

**Comment by tacaswell** [comment]
Well, the minimum numpy is now 1.6 so we should do what ever it was that note was telling us to do.

**Comment by WeatherGod** [comment]
?

**Comment by tacaswell** [comment]
The comment below the commented line says that we are handling `normed` internally until our minimum version of numpy can handle it (which the comment claims in > 1.5).   I think it will take a bit of work (at least for met) to understand how what we are currently doing works to make sure that we can just fall over to numpy.

**Comment by WeatherGod** [comment]
Ah, yeah, that comment (I must be getting blind...). I suspect that it is quite old and came from before the days that stacked histograms and such were possible. I am not sure it is possible to just hand off the responsibility of norming to numpy any more given the current feature set.

**Comment by tacaswell** [comment]
This is passing and not conflicting, can someone push the button on this before it needs another rebase :smiling_imp: .

File: lib/matplotlib/axes/_axes.py (hunk 3 above)
Comment (WeatherGod) [inline lib/matplotlib/axes/_axes.py]: I was waiting for a clearer docstring here because the if-statements are opaque.

**Comment by WeatherGod** [comment]
I do have another concern with respect to generators. Previously, generators just simply didn't work, now they may sort of work (generator gets partly consumed). This muddles the API a bit. Wouldn't it be better to detect that we have a generator and raise an error? We can later consider how to properly support generators without consuming them.

**Comment by tacaswell** [comment]
That is a can of worms of handling I am worried about opening and would rather punt on the whole thing.  This at least does not _break_ any existing user code.

**Comment by WeatherGod** [comment]
isn't it just simply a matter of testing `isinstance(x, collections.Iterator)`? That would be false for lists, numpy arrays and such, but should be true for generators and iterators, I think.

**Comment by tacaswell** [comment]
I am more worried about applying that test uniformly across the library.

**Comment by WeatherGod** [comment]
That would be nice, but it is not the goal that I am seeking. Simply that we add this test where-ever we have code that would potentially consume an iterator. That way, the API hasn't changed. So, that is just two places here in this PR.

**Comment by tacaswell** [comment]
fair enough

**Comment by tacaswell** [comment]
Digging into this a bit (and trying to sort out why `plot` does not have this problem) it turns out that `plot` goes through an entirely different code path by using [`Axis.update_units`](https://github.com/matplotlib/matplotlib/blob/master/lib/matplotlib/axis.py#L1420) which delegates to [`units.Registry.get_convertor`](https://github.com/matplotlib/matplotlib/blob/master/lib/matplotlib/units.py#L125) which has some non-trivial logic and will do very bad things to generators (I think).

This PR is a whack-a-mole fix which will get 1.5.1 out the door.

**Comment by tacaswell** [comment]
The test failure is un-related (due to backport of custom RNG).

**Comment by WeatherGod** [comment]
seems like it, was the baseline images not updated for the new RNG?

**Comment by mdboom** [comment]
> seems like it, was the baseline images not updated for the new RNG?

They were.  This looks like a "only on Travis" sort of failure.

**Comment by WeatherGod** [comment]
That doesn't give me any sort of warm fuzzies...

**Comment by tacaswell** [comment]
I can reproduce it locally (but it passed _before_ I pushed upstream (which I assume means my local install had become screwed up)).

**Comment by WeatherGod** [comment]
I hope this hasn't become another rabbit-infested dragon cave...

**Comment by WeatherGod** [comment]
Looks like you cleared out the rabbits!

**Comment by WeatherGod** [comment]
By moving a bit of the if-statement into a cbook function, the if-statement starts to be a little bit clearer. Probably would be a good idea to have some cbook tests at some point, but that can come later.

**Pull request merged**